src/daft-plan/src/physical_plan.rs

```diff
@@ -220,32 +261,86 @@ impl PhysicalPlan {
+                    stats.upper_bound_bytes = stats
+                        .upper_bound_bytes
+                        .and_then(|st_ub| in_memory_size.map(|ub| st_ub + ub));
+                }
+                stats
             }
-            Self::TabularScan(TabularScan { scan_tasks, .. }) => scan_tasks
-                .iter()
-                .map(|scan_task| scan_task.size_bytes())
-                .sum::<Option<usize>>(),
-            Self::EmptyScan(..) => Some(0),
+            Self::EmptyScan(..) => ApproxStats {
+                lower_bound_rows: 0,
+                upper_bound_rows: Some(0),
+                lower_bound_bytes: 0,
+                upper_bound_bytes: Some(0),
+            },
             // Assume no row/column pruning in cardinality-affecting operations.
             // TODO(Clark): Estimate row/column pruning to get a better size approximation.
-            Self::Filter(Filter { input, .. })
-            | Self::Limit(Limit { input, .. })
-            | Self::Project(Project { input, .. })
-            | Self::MonotonicallyIncreasingId(MonotonicallyIncreasingId { input, .. })
-            | Self::Unpivot(Unpivot { input, .. }) => input.approximate_size_bytes(),
+            Self::Filter(Filter { input, .. }) => {
+                let input_stats = input.approximate_stats();
+                ApproxStats {
+                    lower_bound_rows: 0,
+                    upper_bound_rows: input_stats.upper_bound_rows,
+                    lower_bound_bytes: 0,
+                    upper_bound_bytes: input_stats.upper_bound_bytes,
+                }
+            }
+            Self::Limit(Limit { input, limit, .. }) => {
+                let limit = *limit as usize;
+                let input_stats = input.approximate_stats();
+                let est_bytes_per_row_lower =
+                    input_stats.lower_bound_bytes / (input_stats.lower_bound_rows.max(1));
+                let est_bytes_per_row_upper = input_stats
+                    .upper_bound_bytes
+                    .and_then(|bytes| input_stats.upper_bound_rows.map(|rows| bytes / rows))
+                    .unwrap_or(100);
+                let new_lower_rows = input_stats.lower_bound_rows.min(limit);
+                let new_upper_rows = input_stats
+                    .upper_bound_rows
+                    .map(|ub| ub.min(limit))
+                    .unwrap_or(limit);
+                ApproxStats {
+                    lower_bound_rows: new_lower_rows,
+                    upper_bound_rows: Some(new_upper_rows),
+                    lower_bound_bytes: new_lower_rows * est_bytes_per_row_lower,
+                    upper_bound_bytes: Some(new_upper_rows * est_bytes_per_row_upper),
+                }
+            }
+            Self::Project(Project { input, .. })
+            | Self::MonotonicallyIncreasingId(MonotonicallyIncreasingId { input, .. }) => {
+                // TODO(sammy), we need the schema to estimate the new size per row
```

Comment:
Ah, we can always tweak logical->physical translation to add the schema to the physical `Project` and `MontonicallyIncreasingId` structs for this case! Obviously not a blocking issue, though.